src/Spago/Config/AST.hs

```diff
@@ -0,0 +1,995 @@
+                -- itself is invalid.
+                pure maybeResult
+
+          AtRootExpression key -> do
+            let
+              newLevel = SearchingForField (key :| [])
+            maybeResult <- updateExpr newLevel recordExpr
+            case maybeResult of
+              Just EncounteredEmbed -> do
+                case astMod of
+                  --     ```
+                  --     ./spago.dhall
+                  --       with someField = update
+                  --     ```
+                  --  to
+                  --     ```
+                  --     let __embed = ./spago.dhall
+                  --     in embed
+                  --          with field = update
+                  --          with dependencies = embed.dependencies # ["new"]
+                  --     ```
+                  InsertListText additions -> do
+                    let
+                      varName = "__embed"
+                      var = Dhall.Var (Dhall.V varName 0)
+
+                      -- `let __embed = recordExpr`
+                      binding = Dhall.makeBinding varName recordExpr
+
+                      -- `var.dependencies # ["new"]`
+                      astUpdate =
+                        Dhall.ListAppend (Dhall.Field var (Dhall.makeFieldSelection key))
+                          $ Dhall.ListLit Nothing $ updateListTextByAppending additions Seq.empty
+
+                    pure $ Just $ Updated
+                      $ Dhall.Let binding
+                      -- Note: we can't use `Prefer` here to merge two `With` updates into a single record update
+                      -- (e.g. `foo with key1 = x with key2 = x` -> foo // { key1 = x, key2 = y }` )
+                      -- because we might have a nested selector
+                      -- (e.g. `{ outer = { inner = [] } } with outer.inner = ["foo"]` ).
+                      -- So, we must instead wrap it in another `With`
+                      -- (e.g. `{ outer = { inner = [] }, dependencies = [] } with outer.inner = ["foo"] with dependencies = ["foo"]` ).
+                      $ Dhall.With (Dhall.With var field update) (key :| []) astUpdate
+
+                  --     ```
+                  --     ./spago.dhall
+                  --        with someField = update
+                  --     ```
+                  --  to
+                  --     ```
+                  --     ./spago.dhall
+                  --        with field = update
+                  --        with name = "new"
+                  --     ```
+                  SetText t -> do
+                    pure $ Just $ Updated $ Dhall.With expr (key :| []) t
+
+              Just (VariableName _) -> do
+                -- `The `Just VariableName` can't happen here because this is `AtRootExpression`
+                pure maybeResult
+
+              Just (Updated newUpdate) -> do
+                pure $ Just $ Updated $ Dhall.With recordExpr field newUpdate
```

Comment:
While working on the tests, I found a bug! `newUpdate` should be called `newRecordExpr`. This was likely caused by a copy-paste error.